Add JS to render redirect for new language selection

reginafcompton opened the pull request:
The PR allows the user to select a language using the drop down. 

@jeancochrane I accomplished this entirely with JavaScript. The one difficult thing: we need to know the language selected by the user, even after the page reloads. I did this with `localStorage`, but I did not research the cross-browser reliability of this JS feature....so, now I am wondering if it makes more sense to retrieve the selected language from the URL and then do-the-thing like so. I'd love your input on this one!

jeancochrane commented:
So after a little bit of digging, I think passing links into the views using Django's [request.path](https://docs.djangoproject.com/en/1.11/ref/request-response/#django.http.HttpRequest.path) method and some string formatting might be a better approach, as opposed to relying on Javascript. Do you agree @reginafcompton?

reginafcompton commented:
@jeancochrane - I've refactored this! 

I added a custom filter in `sfm_pc/template_tags/language_url.py`, which accepts the LANGUAGE_CODE as an argument and returns an ordered list (denoting the language to display first in the dropdown).

This filter eliminates a fair amount of JS treachery: we do not use "localStorage," and much of the JS array-reinvention-mumbo-jumbo happens with a couple simple Python methods.

(I did occur to me, about 80% through this, that we might be able to do away with the Select2 in favor of a Django form...that may have been a better solution, but it's only a thought.)

Let me know what you think.

jeancochrane commented on templates/base.html:
```diff
@@ -84,18 +85,31 @@
     });
+
+    // Attach click event to dropdown, which reloads page with new language code 
+    transObj = {
```
Since we use this translation mapper object twice, and I imagine that we might want to edit it in the future, how about we move it to a global variable (like `AVAILABLE_TRANSLATIONS` or something similar) in the settings file and pass it into the view?

reginafcompton replied:
The objects are slightly different - one converts the abbreviation to a full word ('fr' to 'French'), and the other converts the full word to an abbreviation ('French' to 'fr'). I think could consolidate them into one object, e.g., `translation_lookups`...I'll do that and put it in the settings!

jeancochrane replied:
Ooooh OK I see! I missed that this one was performing the opposite lookup.

reginafcompton replied:
...and on second thought, we'll need to set-up a custom context processor to give base.html access to the settings variables, and I am not sure it's worth the effort.

jeancochrane replied:
Fine by me! Maybe if we find ourselves defining the available translations in any more places, we can decide to write the context processor then.

jeancochrane commented on templates/base.html:
```diff
@@ -84,18 +85,31 @@
+        selectedLangSym = '/' + transObj[selectedLang] + '/'
+        // Create new URL with selected language, and redirect to it.
+        currentUrl = window.location.href
+        newUrl = currentUrl.replace('/es/', selectedLangSym).replace('/fr/', selectedLangSym).replace('/en/', selectedLangSym)
```
Nice, I like this.

jeancochrane commented on sfm_pc/templatetags/language_url.py:
```diff
@@ -0,0 +1,26 @@
+      'en': 'English',
+      'es': 'Spanish',
+    }
+
```
Looks like this file might be using two spaces for indexing instead of four.

jeancochrane commented on sfm_pc/templatetags/language_url.py:
```diff
@@ -0,0 +1,26 @@
+
+@register.filter
+def create_select2_data(select_lang):
+  data_array = ["English", "French", "Spanish"]
```
This is a style suggestion more than anything else, but I might instantiate `trans_obj` first (or use a global settings variable, see my comment below) and then create `data_array` like so:

```python
data_array = trans_obj.values()
```

to make the relationship between the two objects explicit.

reginafcompton replied:
Very nice. Ended up with this (thanks, Python 3):

`data_array = list(trans_obj.values())`

jeancochrane replied:
Love Python3!

jeancochrane commented on sfm_pc/templatetags/language_url.py:
```diff
@@ -0,0 +1,26 @@
+
+  full_text = trans_obj[select_lang]
+  data_array.insert(0, full_text)
+  data_array = list(unique_everseen(data_array))
```
Interesting solution! I would have done this with list operations and a little functional programming instead of introducing `more-itertools` as a dependency:

```python
data_array.insert(0, data_array.pop(data_array.index(full_text))
```

But I'm pretty sure these are functionally identical, so it doesn't make much of a difference.

reginafcompton replied:
That's cool! I did know that `pop` could accept an index number as an arg. The only problem if that `.index` finds the first instance of the element, and I'd like to preserve that. 

So, it would be possible to reverse the array, then remove the duplicate, and then reverse it again...what do you think? overkill?

jeancochrane replied:
I think if the element we're after is at the start of the list, everything should still work fine. The code will wind up jumping through a hoop it doesn't actually need to jump through (by removing the first element and adding it to the start of the list, where it already was) but as I understand it that would still produce the right output. Does that make sense?

We could also add a check to avoid that situation altogether:

```python
if data_array[0] != full_text:
    data_array.insert(0, data_array.pop(data_array.index(full_text))
```

reginafcompton replied:
Ah, yes...I like this check...and also, yes, I see how the `pop` can work, here (i.e., we insert after we pop).

Change forthcoming! (And happy to be rid of more-itertools.)

jeancochrane commented on templates/base.html:
```diff
@@ -84,18 +85,31 @@
+    }
+
+    $('.lang-select-box').on('change', function (evt) {
+        selectedLang = $('.lang-select-box').val()
```
I usually mark the end of JS statements with `;`, not sure if you have a different philosophy about that though.

reginafcompton replied:
Oh, yes!

jeancochrane commented:
Also: I like the idea of a Django form, although I can't quite imagine what it would look like. If you want to try for an implementation, go right ahead; otherwise I think this works fine!
